src/database/mod.rs

```diff
@@ -243,10 +253,167 @@ impl Database {
             })
             .collect::<Vec<_>>())
     }
+
+    pub async fn get_root_state(&self, root: &Hash) -> Result<Option<RootItem>, Error> {
+        // This tries really hard to do everything in one query to prevent race
+        // conditions.
+        let query = sqlx::query(
+            r#"SELECT
+                status,
+                COALESCE(
+                    (SELECT r2.created_at
+                    FROM root_history r2
+                    WHERE r2.last_index > r.last_index
+                    ORDER BY r2.last_index
+                    LIMIT 1),
+                    STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW')
+                ) as pending_valid_as_of,
+                COALESCE(
+                    (SELECT r2.mined_at
+                    FROM root_history r2
+                    WHERE r2.last_index > r.last_index AND r2.status = $2
+                    ORDER BY r2.last_index
+                    LIMIT 1),
+                    CASE WHEN r.status = $2 THEN STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW') END
+                ) as mined_valid_as_of
+                FROM root_history r
+                WHERE r.root = $1;
+            "#,
+        )
+        .bind(root)
+        .bind(<&str>::from(Status::Mined));
+
+        let row = self.pool.fetch_optional(query).await?;
+
+        Ok(row.map(|r| {
+            let status = r
+                .get::<&str, _>(0)
+                .parse()
+                .expect("Status is unreadable, database is corrupt");
+            let pending_valid_as_of = r.get::<_, _>(1);
+            let mined_valid_as_of = r.get::<_, _>(2);
+            RootItem {
+                root: *root,
+                status,
+                pending_valid_as_of,
+                mined_valid_as_of,
+            }
+        }))
+    }
+
+    pub async fn insert_pending_root(
+        &self,
+        root: &Hash,
+        last_identity: &Hash,
+        last_index: usize,
+    ) -> Result<(), Error> {
+        let query = sqlx::query(
+            r#"INSERT INTO
+            root_history(root, last_identity, last_index, status, created_at)
+            VALUES($1, $2, $3, $4, STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW'))
+            ON CONFLICT (root) DO NOTHING;"#,
```

Comment:
If there's a conflict, would that potentially mean that we're syncing our internal invalid state to blockchain valid state and therefore we should mark this as pending again?

I know you mentioned that Database is now the source of truth, so maybe it's fine, just want to make sure.